### PR TITLE
Add classifications to clients response TM-1647

### PR DESCRIPTION
**Pull request by scott062**
-Added classifications to clients payload

### CODE REVIEW

File: src/routes.js [hunk 1]
```diff
@@ -180,6 +180,16 @@ var appRouter = function (app) {
       return_code: 0
     })
   })
+
+  app.get('/bidderTrackingPrograms', async function(req, res) {
```
Comment (scott062) [inline src/routes.js]: Adding bidderTrackingPrograms endpoint to this PR - WIP

File: src/services/employees.js [hunk 2]
```diff
@@ -28,6 +28,7 @@ const get_agents = async query => {
     delete emp.manager
     delete emp.manager_id
     delete emp.dob
+    delete emp.classifications
```
Comment (scott062) [inline src/services/employees.js]: @rtirserio I think this is the fix, but agents already returned classifications so I think it has something to do with fetch options

Comment (rtirserio) [reply src/services/employees.js]: the classifications are included in fetch options so this is how to remove it from this response (or create a new query)

File: src/routes.js (hunk 1 above)
Comment (rtirserio) [reply src/routes.js]: is this going to be part of this PR? is this ready for review?

Comment (scott062) [reply src/routes.js]: I added it to this PR - ready for review, but need to fix query returning too many classifications

File: src/services/employees.js [hunk 3]
```diff
@@ -259,4 +261,25 @@ const get_assignments = async query => {
...
+    }).fetchPage()
+    
+
+    return data.serialize().map(classification => {
+      delete classification._pivot_perdet_seq_num
+      delete classification._pivot_td_id
+      return {...classification}
+    })
+  } catch (Error) {
+    return null
```
Comment (rtirserio) [inline src/services/employees.js]: lets log the error here as well 

`console.error(Error)`

File: src/services/employees.js [hunk 4]
```diff
@@ -259,4 +261,25 @@ const get_assignments = async query => {
...
+        qb.join('employees_classifications', 'employees_classifications.td_id', 'classifications.td_id')
+        qb.where('employees_classifications.perdet_seq_num', perdet_seq_num)
+      }
+    }).fetchPage()
+    
+
+    return data.serialize().map(classification => {
+      delete classification._pivot_perdet_seq_num
+      delete classification._pivot_td_id
+      return {...classification}
```
Comment (rtirserio) [inline src/services/employees.js]: no need to explode the object, just return the classification

Comment (scott062) [reply src/services/employees.js]: Fixed! Thanks

File: src/services/employees.js (hunk 3 above)
Comment (scott062) [reply src/services/employees.js]: Done!

**Comment by scott062** [comment]
Added tests - ready for final review

**Comment by mjoyce91** [comment]
@scott062 - if one classification is returned, then it should be returned as a single object (to match fsbid)

**Comment by scott062** [comment]
@mjoyce91 For /bidderTrackingPrograms, does it always returns an object for single and multiple classifications?

**Comment by mjoyce91** [comment]
@scott062 - that one actually returns an array regardless

**Comment by scott062** [comment]
Hopefully done - Criteria

- Agents payload does not return classifications
- Clients payload does return classifications
1) returns single obj for single classification
2) return array of objs for multiple classification
3) return NOTHING for client with no classifications (field is deleted from payload) 
- bidderTrackingProgram returns just classifications for single client by perdet always an array of classification objs

File: src/services/employees.js [hunk 5]
```diff
@@ -91,6 +91,13 @@ const get_clients = async query => {
         },
       }
     }
+    // This returns the payload without any classifications field when the client has none to mimic fsbid
+    if (res.employee.classifications.length < 1) {
+      delete res.employee.classifications
+      return res 
+    } else {
+      return res
+    }
```
Comment (mjoyce91) [inline src/services/employees.js]: Can be simplified to:

```js
if (res.employee.classifications.length < 1) {
      delete res.employee.classifications
}
return res
```

Comment (scott062) [reply src/services/employees.js]: Thanks Mike! 💯

File: src/services/employees.js [hunk 6]
```diff
@@ -74,6 +76,7 @@ const get_clients = async query => {
         ...personSkills(emp.skills),
         per_pay_plan_code: "",
         per_tenure_code: "",
+        classifications: classifications.length === 1 ? classifications[0] : classifications,
```
Comment (mjoyce91) [inline src/services/employees.js]: 😢

File: src/services/employees.js [hunk 7]
```diff
@@ -268,4 +276,25 @@ const get_assignments = async query => {
...
+      if (perdet_seq_num) {
+        qb.join('employees_classifications', 'employees_classifications.td_id', 'classifications.td_id')
+        qb.where('employees_classifications.perdet_seq_num', perdet_seq_num)
+      }
+    }).fetchPage()
+    
+    return data.serialize().map(classification => {
+      delete classification._pivot_perdet_seq_num
+      delete classification._pivot_td_id
+      return classification
```
Comment (mjoyce91) [inline src/services/employees.js]: These don't appear to be getting deleted. I would install lodash and do:

```js
import { omit } from 'lodash';
...
return omit(classification, ['_pivot_perdet_seq_num', '_pivot_td_id']);
```

You can also do this in plain JS with:

```
const { _pivot_perdet_seq_num, _pivot_td_id, ...myFilteredObject } = classification;
return classification;
```

Comment (mjoyce91) [reply src/services/employees.js]: In general I just avoid using `delete` as much as possible

File: src/services/employees.js (hunk 6 above)
Comment (scott062) [reply src/services/employees.js]: 😭

File: src/services/employees.js [hunk 8]
```diff
@@ -102,10 +102,21 @@ const get_clients = async query => {
...
+    } else if (res.employee.classifications.length > 1) {
+      // Classifications as array
+      res.employee.classifications.forEach((classification, i) => {
+        const { _pivot_perdet_seq_num, _pivot_td_id, ...filteredClassification } = classification
+        res.employee.classifications[i] = filteredClassification
+      })
+    } else {
+      // Single classification as object
+      const { _pivot_perdet_seq_num, _pivot_td_id, ...filteredClassifications } = classifications
+      res.employee.classifications = filteredClassifications
```
Comment (mjoyce91) [inline src/services/employees.js]: This is getting returned with an extra nested object `"0"`:

```json
"classifications": {
          "0": {
            "td_id": 135,
            "tp_code": "P",
            "tp_descr_txt": "Pickering/Rangel Fellows",
            "disabled_ind": "N",
            "_pivot_perdet_seq_num": 16,
            "_pivot_td_id": 135
          }
        },
```

Comment (mjoyce91) [reply src/services/employees.js]: And it includes the _pivot properties

**Pull request merged**